server/src/main/java/com/tencent/rss/server/ShuffleTaskManager.java

```diff
@@ -207,12 +208,32 @@ public void addFinishedBlockIds(
       shuffleIdToPartitions.putIfAbsent(shuffleId, blockIds);
     }
     Roaring64NavigableMap[] blockIds = shuffleIdToPartitions.get(shuffleId);
+
+    // This method will be called from many threads, it's important to keep lock in a reasonable way
+    // Here is the strategy:
+    // 1. get lock of bitmap
+    // 2. update the bitmap
+    // 3. do 1 & 2 again for another bitmap
+    // Todo: High latency with current implementation for the lock competition
+    List<List<long[]>> entryList = Lists.newArrayList();
+    for (int i = 0; i < bitmapNum; i++) {
+      entryList.add(i, Lists.newArrayList());
+    }
+
+    // prepare the data for each bitmap
     for (Map.Entry<Integer, long[]> entry : partitionToBlockIds.entrySet()) {
       Integer partitionId = entry.getKey();
-      Roaring64NavigableMap bitmap = blockIds[partitionId % bitmapNum];
+      entryList.get(partitionId % bitmapNum).add(entry.getValue());
+    }
+
+    for (int i = 0; i < bitmapNum; i++) {
+      Roaring64NavigableMap bitmap = blockIds[i];
+      List<long[]> blockIdList = entryList.get(i);
       synchronized (bitmap) {
```

Comment:
To be honesty, I don't ensure that these will solve the problem. We should some mircro benchmark at least.

Reply:
Yes，according to the simulator test, the improvement can get 4X performance

Reply:
Can you describe some details about the simulator test?